src/main/java/com/amazon/aws/iot/greengrass/semver/SemVer.java

```diff
@@ -0,0 +1,340 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.MatchResult;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class SemVer implements Comparable<SemVer> {
+    private static final Pattern IS_NUMERIC = Pattern.compile("^[0-9]+$");
+    private final String raw;
+    @Getter
+    private String version;
+    public final int major;
+    public final int minor;
+    public final int patch;
+    public final List<Object> prerelease;
+    private final List<String> build;
+
+    SemVer(int major, int minor, int patch, List<Object> prerelease, List<String> build) {
+        this.major = major;
+        this.minor = minor;
+        this.patch = patch;
+        this.prerelease = prerelease;
+        this.build = build;
+        this.raw = this.format();
+        this.version = this.raw;
+    }
+
+    SemVer(SemVer other) {
+        this(other.major, other.minor, other.patch, new ArrayList<>(other.prerelease), new ArrayList<>(other.build));
+    }
+
+    // loose and includePrerelease are false
+    public SemVer(String version) {
+        if (version == null) {
+            throw new IllegalArgumentException("Version \"null\" is not valid");
+        }
+        Matcher matcher = SemVerRegexes.re.get(SemVerRegexes.t.get("FULL")).matcher(version.trim());
+        if (!matcher.find()) {
+            throw new IllegalArgumentException("Version \"" + version + "\" is not valid");
+        }
+        MatchResult m = matcher.toMatchResult();
+
+        this.raw = version;
+
+        // these are actually numbers
+        this.major = Integer.parseInt(m.group(1));
+        this.minor = Integer.parseInt(m.group(2));
+        this.patch = Integer.parseInt(m.group(3));
+
+        if (this.major < 0) {
+            throw new IllegalArgumentException("Invalid major version");
+        }
+
+        if (this.minor < 0) {
+            throw new IllegalArgumentException("Invalid minor version");
+        }
+
+        if (this.patch < 0) {
+            throw new IllegalArgumentException("Invalid patch version");
+        }
+
+        // numberify any prerelease numeric ids
+        if (m.group(4) != null) {
+            this.prerelease = Arrays.stream(m.group(4).split("\\.", -1)).map((String id) -> {
+                if (IS_NUMERIC.matcher(id).find()) {
+                    int num = Integer.parseInt(id);
+                    if (num >= 0) {
+                        return num;
+                    }
+                }
+                return id;
+            }).collect(Collectors.toList());
+        } else {
+            this.prerelease = new ArrayList<>();
+        }
+
+        this.build = m.group(5) == null ? new ArrayList<>() : Arrays.asList(m.group(5).split("\\.", -1));
+        this.version = this.format();
+    }
+
+    protected String format() {
+        String v = String.format("%s.%s.%s", this.major, this.minor, this.patch);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: patch, major, minor. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 102. Consider formatting this data with &#39;%d&#39; instead.